### PR TITLE
Reorganize project runs

### PR DESCRIPTION
**Pull request by volkangurel**
Several important pieces of cleanup:
- Remove deprecated "Raw Datasets"
- Remove hyper parameter tuning
- Refactor how projects and runs work
- Deduplicate Function definitions, simplify logic
- Allow dependencies to be specified by strings

Fixes LAY-3126

### CODE REVIEW

**Review by cihati — COMMENTED** [review]
I wish this was divided commit-by-commit, would make review a lot easier :/

I'm taking a detailed look. Also, please remove:

```layer/exceptions/exceptions.py:class ProjectHPTExecutionException(ProjectExecutionException):```

**Review by cihati — COMMENTED** [review]
Maybe get rid of RawDataset fully, e.g.

```
layer/clients/data_catalog.py:    RawDataset,
layer/clients/data_catalog.py:            raw_text = str.format("Dataset build failed with {}", error)
layer/clients/data_catalog.py:                info=(raw_text[: (max_error_length - len(placeholder))] + placeholder)
layer/clients/data_catalog.py:                if len(raw_text) > max_error_length
layer/clients/data_catalog.py:                else raw_text
layer/clients/data_catalog.py:        return RawDataset(
layer/contracts/datasets.py:        :param sampler: defines the strategy to draw samples from the dataset. Can be any Iterable with __len__ implemented. If specified, shuffle must not be specified.
layer/contracts/datasets.py:class RawDataset(Dataset):
layer/contracts/datasets.py:    def with_table_name(self, name: str) -> "RawDataset":
layer/contracts/datasets.py:    def with_metadata(self, metadata: Mapping[str, str]) -> "RawDataset":
layer/contracts/datasets.py:    def with_project_name(self, project_name: str) -> "RawDataset":
layer/contracts/entities.py:    RAW_DATASET = "raw_dataset"
```

File: layer/contracts/runs.py [hunk 1]
```diff
@@ -1,15 +1,227 @@
...
+        self.name = name
+        fabric = layer_settings.get_fabric()
+        if fabric is None:
+            fabric = Fabric.F_LOCAL.value
+        self._fabric = fabric
+        self._dependencies = layer_settings.get_dependencies()
+        self.pip_packages = layer_settings.get_pip_packages()
+        self.pip_requirements_file = layer_settings.get_pip_requirements_file()
+
+        self.func_source = inspect.getsource(self.func)
```
Comment (cihati) [inline layer/contracts/runs.py:106]: nit: I'd name this `func_source_code`

File: layer/contracts/runs.py [hunk 2]
```diff
@@ -1,15 +1,227 @@
...
+    @property
+    def dependencies(self) -> List[AssetPath]:
+        return [d.with_project_name(self.project_name) for d in self._dependencies]
+
+    @property
+    def entrypoint(self) -> str:
+        return f"{self.name}.pkl"
+
+    @property
+    def entity_path(self) -> Path:
```
Comment (cihati) [inline layer/contracts/runs.py:137]: we should rename this, I think. It feels too ambiguous, especially given we also have `asset_path` too.

File: layer/clients/model_catalog.py [hunk 3]
```diff
@@ -134,29 +98,72 @@ def create_model_version(
...
         load_response = self._service.LoadModelTrainDataByPath(
             LoadModelTrainDataByPathRequest(path=path),
         )
-        return ModelDefinition(
-            name=path,
-            train_id=load_response.id,
-            proto_flavor=load_response.flavor,
-            s3_path=load_response.s3_path,
-            credentials=load_response.credentials,
+        flavor = get_flavor_for_proto(load_response.flavor)
```
Comment (cihati) [inline layer/clients/model_catalog.py:153]: This function is only used here -- perhaps move it here and move the exception raise logic into it too?

File: layer/clients/model_catalog.py [hunk 4]
```diff
@@ -170,53 +177,83 @@ def create_model_train_from_version_id(
...
-        state: Optional[ResourceTransferState] = None,
-    ) -> Any:
+    ) -> ModelArtifact:
         """
-        Loads a model from the model catalog
+        Loads a model artifact from the model catalog
 
-        :param model_definition: model definition
-        :return: a model object
+        :param model: the model
```
Comment (cihati) [inline layer/clients/model_catalog.py:189]: might as well update the docs for other params too = )

**Comment by cihati** [comment]
Also, I'd nuke `model-training/python-executor/pytrain/hpt_*` stuff first, before merging this PR

File: layer/contracts/models.py [hunk 5]
```diff
@@ -225,11 +133,6 @@ def with_project_name(self, project_name: str) -> "Model":
         new_model._update_with(new_asset)
         return new_model
 
-    def with_language_version(self, language_version: Tuple[int, int, int]) -> "Model":
```
Comment (cihati) [inline layer/contracts/models.py:228]: you can delete this from datasets.py too

**Comment by cihati** [comment]
The PR is extremely large, and is not even split into commits. It makes the review extremely hard. Can you please chop it up and send several PRs please? Also, is it possible to use this repo directly and create a branch (vs forking)?

**Comment by volkangurel** [comment]
> Maybe get rid of RawDataset fully, e.g.
> 
> ```
> layer/clients/data_catalog.py:    RawDataset,
> layer/clients/data_catalog.py:            raw_text = str.format("Dataset build failed with {}", error)
> layer/clients/data_catalog.py:                info=(raw_text[: (max_error_length - len(placeholder))] + placeholder)
> layer/clients/data_catalog.py:                if len(raw_text) > max_error_length
> layer/clients/data_catalog.py:                else raw_text
> layer/clients/data_catalog.py:        return RawDataset(
> layer/contracts/datasets.py:        :param sampler: defines the strategy to draw samples from the dataset. Can be any Iterable with __len__ implemented. If specified, shuffle must not be specified.
> layer/contracts/datasets.py:class RawDataset(Dataset):
> layer/contracts/datasets.py:    def with_table_name(self, name: str) -> "RawDataset":
> layer/contracts/datasets.py:    def with_metadata(self, metadata: Mapping[str, str]) -> "RawDataset":
> layer/contracts/datasets.py:    def with_project_name(self, project_name: str) -> "RawDataset":
> layer/contracts/entities.py:    RAW_DATASET = "raw_dataset"
> ```

Good callout, I wanted to limit the size of the PR since it was already quite large, but I can add that here.

**Comment by volkangurel** [comment]
> The PR is extremely large, and is not even split into commits. It makes the review extremely hard. Can you please chop it up and send several PRs please? Also, is it possible to use this repo directly and create a branch (vs forking)?

I'm sure I can work to split this up, but unfortunately the projects, runs, HPT, etc were so deeply intertwined that it would be very difficult to draw a clear line. This is one problem that this PR is attempting to solve - to remove decoupling from all these components and make it easier to reason about while reducing the number of lines of code.

File: layer/projects/execution_planner.py [hunk 6]
```diff
@@ -131,77 +99,112 @@ def build_execution_plan(
...
-        ]
-        res[level] = [
-            PlanNode(vertex[0], vertex[1], _graph.nodes[vertex]["entity"])
-            for vertex in zero_in_degree
-        ]
-        _graph.remove_nodes_from(zero_in_degree)
-        level = level + 1
-    return res
+def check_entity_dependencies(definitions: Sequence[FunctionDefinition]) -> None:
+    _build_directed_acyclic_graph(definitions)
```
Comment (cihati) [inline layer/projects/execution_planner.py:103]: This feels a bit weird. Not necessary to address in this PR, but maybe for followups

File: layer/tracker/local_execution_project_progress_tracker.py [hunk 7]
```diff
@@ -141,34 +142,31 @@ def _update_entity(  # noqa: C901
             self._progress.stop_task(task_id)
             self._progress.update(task_id, completed=0, description="error")
 
-    def mark_derived_dataset_saved(self, name: str, *, id_: uuid.UUID) -> None:
+    def _get_url(self, asset_type: AssetType, name: str) -> URL:
```
Comment (cihati) [inline layer/tracker/local_execution_project_progress_tracker.py:145]: For upcoming PR: move this to a superclass?

**Review by cihati — APPROVED** [review]
Another idea for future:
* AssetType vs EntityType deduplication

**Comment by cihati** [comment]
@volkangurel this is very hairy, thanks for getting into this. I agree that it's not possible to make this many small PRs, but I'd personally at least do the following (for instance, commit-by commit):

* Remove raw dataset logic
* Remove HPT logic
* Do renames
* Do the rest

**Comment by maasg** [comment]
I didn't want to remove `RawDataset` right away b/c it's the only means we have to register an external dataset.
We will probably need it when/if we want to register a `dbt` pipeline and its sources.

File: layer/contracts/runs.py (hunk 2 above)
Comment (volkangurel) [reply layer/contracts/runs.py:144]: I will do this in a separate PR for refactoring assets and entities.

File: layer/tracker/local_execution_project_progress_tracker.py (hunk 7 above)
Comment (volkangurel) [reply layer/tracker/local_execution_project_progress_tracker.py:145]: 👍

File: layer/projects/execution_planner.py (hunk 6 above)
Comment (volkangurel) [reply layer/projects/execution_planner.py:103]: Weird in what way?

**Comment by volkangurel** [comment]
> I didn't want to remove `RawDataset` right away b/c it's the only means we have to register an external dataset. We will probably need it when/if we want to register a `dbt` pipeline and its sources.

I think it's best to rethink what metadata we'll need for DBT datasets rather than reuse raw datasets. We can always look at the code here and pull the pieces that we need, but what was in the SDK previously wasn't using RawDatasets correctly IMHO.

File: layer/contracts/models.py (hunk 5 above)
Comment (volkangurel) [reply layer/contracts/models.py:228]: saving this for future cleanup

File: layer/clients/model_catalog.py (hunk 4 above)
Comment (volkangurel) [reply layer/clients/model_catalog.py:189]: saving this for future cleanup of docstrings

File: layer/contracts/runs.py (hunk 1 above)
Comment (volkangurel) [reply layer/contracts/runs.py:113]: saving this for future cleanup

File: layer/clients/model_catalog.py (hunk 3 above)
Comment (volkangurel) [reply layer/clients/model_catalog.py:153]: saving this for future cleanup

**Pull request merged**